### PR TITLE
Add MinigameStates

### PR DESCRIPTION
**Pull request by Flibio**
Adds MinigameStates. A Minigame defines the MinigameStates, and the Arena consumes them. The Arena progresses state by state in chronological order, until it reaches the final state. The Arena will then reset and return to the original state.

### CODE REVIEW

File: src/main/java/io/github/minigamecore/api/minigame/arena/Arena.java [hunk 1]
```diff
@@ -66,4 +67,21 @@
      */
     List<UUID> getPlayers();
```
Comment (kinggoesgaming) [inline src/main/java/io/github/minigamecore/api/minigame/arena/Arena.java]: List -> Collection

File: src/main/java/io/github/minigamecore/api/minigame/Minigame.java [hunk 2]
```diff
@@ -28,10 +28,39 @@
...
  */
 public interface Minigame extends DataManager {
 
+    /**
+     * Gets all possible states of the minigame. The states are ordered
+     * ascending by their key.
+     * 
+     * @return All possible states of the minigame
+     */
+    TreeMap<Integer, MinigameState> getStates();
```
Comment (kinggoesgaming) [inline src/main/java/io/github/minigamecore/api/minigame/Minigame.java]: any reason for this to be a TreeMap instead of Map?

Comment (kinggoesgaming) [inline src/main/java/io/github/minigamecore/api/minigame/Minigame.java]: TreeMap -> SortedMap

File: src/main/java/io/github/minigamecore/api/minigame/Minigame.java [hunk 3]
```diff
@@ -28,10 +28,39 @@
...
+
+    /**
+     * Gets the intial state of the minigame. May return null if no states are
+     * present.
+     * 
+     * @return The initial state of the minigame.
+     */
+    default MinigameState getInitialState() {
+        return getStates().firstEntry().getValue();
```
Comment (kinggoesgaming) [inline src/main/java/io/github/minigamecore/api/minigame/Minigame.java]: `return getStates().get(getStates().firstKey());`

File: src/main/java/io/github/minigamecore/api/minigame/Minigame.java [hunk 4]
```diff
@@ -28,10 +28,39 @@
...
+    }
+
+    /**
+     * Gets the final state of the minigame. May return null if no states are
+     * present.
+     * 
+     * @return The final state of the minigame.
+     */
+    default MinigameState getFinalState() {
+        return getStates().lastEntry().getValue();
```
Comment (kinggoesgaming) [inline src/main/java/io/github/minigamecore/api/minigame/Minigame.java]: `return getStates().get(getStates().lastKey());`

File: src/main/java/io/github/minigamecore/api/minigame/Minigame.java [hunk 5]
```diff
@@ -28,10 +28,39 @@
...
+    }
```
Comment (kinggoesgaming) [inline src/main/java/io/github/minigamecore/api/minigame/Minigame.java]: newline

File: src/main/java/io/github/minigamecore/api/minigame/arena/Arena.java [hunk 6]
```diff
@@ -66,4 +67,21 @@
...
+     * 
+     * @return The current {@link MinigameState} of this arena.
+     */
+    MinigameState getState();
+
+    /**
+     * Cycles the arena to the next {@link MinigameState}.
+     */
+    void nextState();
+
```
Comment (kinggoesgaming) [inline src/main/java/io/github/minigamecore/api/minigame/arena/Arena.java]: `void previousState()`?

File: src/main/java/io/github/minigamecore/api/minigame/Minigame.java (hunk 3 above)
Comment (Flibio) [inline src/main/java/io/github/minigamecore/api/minigame/Minigame.java]: @kinggoesgaming: What is the benefit of your method over the current code?

File: src/main/java/io/github/minigamecore/api/minigame/Minigame.java (hunk 4 above)
Comment (Flibio) [inline src/main/java/io/github/minigamecore/api/minigame/Minigame.java]: @kinggoesgaming: What is the benefit of your method over the current code?

File: src/main/java/io/github/minigamecore/api/minigame/MinigameState.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+     * @return Yhe runnable the state will use on execution.
+     */
+    Runnable getRunnable();
+
+    /**
+     * Executes the runnable of this state.
+     */
+    default void execute() {
+        getRunnable().run();
+    }
```
Comment (kinggoesgaming) [inline src/main/java/io/github/minigamecore/api/minigame/MinigameState.java]: newline

File: src/main/java/io/github/minigamecore/api/minigame/MinigameState.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Represents a state of a {@link Minigame}.
+ */
+public interface MinigameState extends CatalogType {
+
+    /**
+     * Gets the runnable the state will use on execution.
+     * 
+     * @return Yhe runnable the state will use on execution.
```
Comment (Flibio) [inline src/main/java/io/github/minigamecore/api/minigame/MinigameState.java]: `The` not `Yhe`.

**Comment by Flibio** [comment]
cb5b4080f3c9556e722080b81fbcf340c54b1620

**Pull request closed without merging**